AM6x A53 pinctrl fix

gramsay0 opened the pull request:
This PR is intended to fix any mistakes made in https://github.com/zephyrproject-rtos/zephyr/pull/55245.
See https://github.com/zephyrproject-rtos/zephyr/pull/56565 for further discussion.

Main changes (so far):

- Remove direct usage of `pinctrl-0` props in drivers
- Convert pinctrl driver to multi-instance

gramsay0 committed:
drivers: serial: Fix pinctrl usage in NS16550 driver

pinctrl-0 property should not be directly referenced in this driver

Signed-off-by: Grant Ramsay <gramsay@enphaseenergy.com>

gramsay0 committed:
drivers: pinctrl: Convert TI K3 pinctrl driver to multi-instance

K3 devices have multiple pinctrl peripherals

Signed-off-by: Grant Ramsay <gramsay@enphaseenergy.com>

gramsay0 committed:
soc: arm64: ti_sitara: Add MCU domain pinctrl for TI AM62X A53 SoC

Allows accessing the MCU (Cortex-M4) domain pinctrl
peripheral from the MAIN (Cortex-A53) domain

Signed-off-by: Grant Ramsay <gramsay@enphaseenergy.com>

gmarull commented:
See my comment here as well https://github.com/zephyrproject-rtos/zephyr/pull/56565#issuecomment-1505049967

gramsay0 commented on drivers/pinctrl/pinctrl_ti_k3.c:
```diff
@@ -40,11 +35,20 @@ static int pinctrl_ti_k3_init(const struct device *dev)
+		CONFIG_KERNEL_INIT_PRIORITY_DEFAULT,    \
+		NULL);
+
+DT_INST_FOREACH_STATUS_OKAY(PINCTRL_TI_K3_INIT)
```
I think in this case it makes sense for pinctrl to be multi-instance.
K3 devices are basically 3 MCUs (domains) spliced together:

- Main (Cortex-A)
- MCU (Cortex-M/Cortex-R)
- Wakeup (Cortex-R)

There is a crossbar/interconnect (CBASS) that allows inter-domain access to peripherals.

For AM62 devices the main and MCU domains both have a pinctrl "peripheral" (a memory block associated with this functionality). Keeping these separate makes it obvious what pins are being referenced.

e.g. Linux has two pinctrl instances for AM62:

	main_pmx0: pinctrl@f4000 {
		compatible = "pinctrl-single";
		reg = <0x00 0xf4000 0x00 0x2ac>;
		#pinctrl-cells = <1>;
		pinctrl-single,register-width = <32>;
		pinctrl-single,function-mask = <0xffffffff>;
	};
.

	mcu_pmx0: pinctrl@4084000 {
		compatible = "pinctrl-single";
		reg = <0x00 0x04084000 0x00 0x88>;
		#pinctrl-cells = <1>;
		pinctrl-single,register-width = <32>;
		pinctrl-single,function-mask = <0xffffffff>;
	};

gmarull replied:
From a Zephyr perspective, you can only have a single pin controller (remember pin control calls do not take any instance), whether pinctrl driver manages more than a single one internally is up to the implementation. Also, I assume that each core will run its own firmware image, so in theory, each instance will have pin control facilities at some level.

gmarull replied:
Btw, it looks like AM62 uses these bindings: https://www.kernel.org/doc/Documentation/devicetree/bindings/pinctrl/pinctrl-single.txt. I'd suggest spending some time to align with Linux.

gramsay0 replied:
> remember pin control calls do not take any instance

The pin control APIs take a `struct pinctrl_dev_config` which contains a SoC specific `pinctrl_soc_pin_t`.
In this PR I added the pinctrl device instance to `pinctrl_soc_pin_t`.

    #define TI_SITARA_DT_PIN(node_id)                  \
    {                                                  \
        .dev = DEVICE_DT_GET(DT_PARENT(node_id)),      \
        ...
Although this is slightly hidden, I believe it is still multi-instance.

> Btw, it looks like AM62 uses these bindings: https://www.kernel.org/doc/Documentation/devicetree/bindings/pinctrl/pinctrl-single.txt. I'd suggest spending some time to align with Linux.

Ok, will do

gramsay0 replied:
@gmarull I've had a look at making the device tree match Linux.

From what I can tell I only need to support "pin**ctrl**-single" not "pin**conf**-single" for k3 devices (which simplifies a lot).

Although one issue I'm having is trying to make the Linux style pin definition work:

	uart0_pins_default: uart0_pins_default {
		pinctrl-single,pins =
			AM62X_IOPAD(0x1c8, PIN_INPUT, MUX_MODE0)
			AM62X_IOPAD(0x1cc, PIN_OUTPUT, MUX_MODE0)
		>;
	};

evaluates to:

    #define DT_N_S_pinctrl_f4000_S_uart0_pins_default_P_pinctrl_single_pins {456 /* 0x1c8 */, 327680 /* 0x50000 */, 460 /* 0x1cc */, 65536 /* 0x10000 */}

Which is two sets of pinctrl combined. These need to be somehow sliced to `{456, 327680}` and `{460, 65536}` to fit into this macro:

	#define TI_SITARA_DT_PIN(node_id)						\
		{									\
			.dev = DEVICE_DT_GET(DT_PARENT(node_id)),			\
			.offset = DT_PROP_BY_IDX(node_id, pinctrl_single_pins, 0),	\
			.value = DT_PROP_BY_IDX(node_id, pinctrl_single_pins, 1)	\
		},
		
(The size of the slice comes from `#pinctrl-cells = <1>;` + 1 = 2 in this case).
I couldn't find any macros available in zephyr to do this.

Any ideas on how the above could be accomplished?
How important is it to match the Linux DT exactly?

gmarull replied:
> @gmarull I've had a look at making the device tree match Linux.
> 
> From what I can tell I only need to support "pin**ctrl**-single" not "pin**conf**-single" for k3 devices (which simplifies a lot).
> 
> Although one issue I'm having is trying to make the Linux style pin definition work:
> 
> ```
> uart0_pins_default: uart0_pins_default {
> 	pinctrl-single,pins =
> 		AM62X_IOPAD(0x1c8, PIN_INPUT, MUX_MODE0)
> 		AM62X_IOPAD(0x1cc, PIN_OUTPUT, MUX_MODE0)
> 	>;
> };
> ```
> 
> evaluates to:
> 
> ```
> #define DT_N_S_pinctrl_f4000_S_uart0_pins_default_P_pinctrl_single_pins {456 /* 0x1c8 */, 327680 /* 0x50000 */, 460 /* 0x1cc */, 65536 /* 0x10000 */}

These all look like 32-bit values, ie 0x1c850000 or 0x1cc10000, so why are you expanding them to an array like this? I think this is the problem. 

Btw, we should accelerate these changes or revert the original implementation. Please prioritize before the next release.

gramsay0 replied:
The idea of pinctrl-single seems to be a generic driver that could be used across many different SoCs/vendors.

I think it's just by chance that these values would fit into a 32bit number. The proper way to handle this array is to use the `#pinctrl-cells` node to know how many array items are used per pin. I think that may require modifications to the python device tree code.

I've pushed the changes for where I got up to, it still has the issue I mentioned above. Let me know if you have any ideas on how to fix that.

I can add small fixes, but don't have time to dive into python device tree changes.
If you think this PR still needs a lot of work, then I'm okay for you to revert the original PR

gramsay0 committed:
drivers: pinctrl: Convert TI K3 pinctrl driver to pinctrl-single

pinctrl-single is a generic pinctrl driver used in Linux.
One-register-per-pin type device tree based pinctrl driver.

Signed-off-by: Grant Ramsay <gramsay@enphaseenergy.com>

gmarull reviewed (changes requested):
I think this is in better shape now, but we are not yet aligned with Linux bindings. Please make the necessary changes to maintain compatibility.

gramsay0 committed:
wip: Allow iterating pinctrl cells array by pin sized chunks

Add support for DT_FOREACH_PINCTRL_CELLS_GROUP() macro.
This macro can be used on pinctrl nodes whose parent device
contains a "#pinctrl-cells" property.

Signed-off-by: Grant Ramsay <gramsay@enphaseenergy.com>

gramsay0 commented:
@gmarull I've added a WIP commit that allows slicing the `pinctrl-single,pins` array to take `#pinctrl-cells + 1` values at a time.

e.g. with `#pinctrl-cells = <1>;`

	pinctrl-single,pins = <
		0x1c8 0x50000
		0x1cc 0x10000
	>;

This macro gives two indexes at a time:

	DT_FOREACH_PINCTRL_CELLS_GROUP(node_id, prop, fn)
		- fn(node_id, prop, 0, 1)
		- fn(node_id, prop, 2, 3)

and with `#pinctrl-cells = <2>;`

	pinctrl-single,bits = <
		0x00 0x11111111 0xffffffff
		0x04 0x00011000 0x000ff000
	>;

The macro gives three indexes at a time:

	DT_FOREACH_PINCTRL_CELLS_GROUP(node_id, prop, fn)
		- fn(node_id, prop, 0, 1, 2)
		- fn(node_id, prop, 3, 4, 5)

If you think this is a reasonable approach I will tidy it up and add some unit tests

gmarull commented:
Thanks for the continued work @gramsay0. I think that the `pinctrl-cells` usage is the correct approach (also used in Linux). I wonder if the `DT_PHA_BY_IDX` API is not enough (provided you also add in bindings the `pinctrl-cells` specifier). Have you tried?

gramsay0 commented:
> Thanks for the continued work @gramsay0. I think that the `pinctrl-cells` usage is the correct approach (also used in Linux). I wonder if the `DT_PHA_BY_IDX` API is not enough (provided you also add in bindings the `pinctrl-cells` specifier). Have you tried?

@gmarull It's not easy to split the array into groups using preprocessor macros.

Without touching the python device tree code I could hard code something like this into "pinctrl_soc.h":

	#define FOREACH_PINCTRL_CELLS_GROUP(node_id, prop, fn)                                \
		IF_ENABLED(DT_PROP_HAS_IDX(node_id, prop, 0),  ( fn(node_id, prop, 0, 1)   )) \
		IF_ENABLED(DT_PROP_HAS_IDX(node_id, prop, 2),  ( fn(node_id, prop, 2, 3)   )) \
		IF_ENABLED(DT_PROP_HAS_IDX(node_id, prop, 4),  ( fn(node_id, prop, 4, 5)   )) \
		IF_ENABLED(DT_PROP_HAS_IDX(node_id, prop, 6),  ( fn(node_id, prop, 6, 7)   )) \
		IF_ENABLED(DT_PROP_HAS_IDX(node_id, prop, 8),  ( fn(node_id, prop, 8, 9)   )) \
		IF_ENABLED(DT_PROP_HAS_IDX(node_id, prop, 10), ( fn(node_id, prop, 10, 11) )) \
		IF_ENABLED(DT_PROP_HAS_IDX(node_id, prop, 12), ( fn(node_id, prop, 12, 13) )) \
		IF_ENABLED(DT_PROP_HAS_IDX(node_id, prop, 14), ( fn(node_id, prop, 14, 15) )) \
		IF_ENABLED(DT_PROP_HAS_IDX(node_id, prop, 16), ( BUILD_ASSERT(false);      ))

This would work for up to 8 pins and assumes `#pinctrl-cells` is set to 1.
Maybe that's good enough for now?

`#pinctrl-cells` already has bindings, but anything that starts with `#` or ends with `-map` does not generate defines/macros in the device tree header
